src/NUnitFramework/framework/Internal/Results/TestResult.cs

```diff
@@ -480,23 +471,54 @@ public void RecordException(Exception ex, FailureSite site)
         /// <param name="ex">The Exception to be recorded</param>
         public void RecordTearDownException(Exception ex)
         {
-            var result = new ExceptionResult(ex, FailureSite.TearDown);
+            RecordException(ex, FailureSite.TearDown);
+        }
+
+        /// <summary>
+        /// Set the test result based on the type of exception thrown
+        /// </summary>
+        /// <param name="ex">The exception that was thrown</param>
+        /// <param name="site">The FailureSite to use in the result</param>
+        public void RecordException(Exception ex, FailureSite site)
+        {
+            var result = new ExceptionResult(ex, site);
 
-            ResultState resultState = ResultState == ResultState.Cancelled
-                ? ResultState.Cancelled
-                : ResultState.Error;
-            if (Test.IsSuite)
-                resultState = resultState.WithSite(FailureSite.TearDown);
+            if (ex is NUnitException nUnitException && nUnitException.InnerException is ResultStateException ||
+                ex is ResultStateException)
```

Comment:
```suggestion
              
```

Reply:
If the user executes `Assert.Fail/Ignore/Success` in a SetUp or TearDown it will get here with a (derived) `ResultStateException`